executor/compiler.go

```diff
@@ -42,18 +42,14 @@ func (c *Compiler) Compile(ctx context.Context, stmtNode ast.StmtNode) (*ExecStm
 		return nil, errors.Trace(err)
 	}
 
-	readOnlyCheckStmt := stmtNode
-	if checkPlan, ok := finalPlan.(*plan.Execute); ok {
-		readOnlyCheckStmt = checkPlan.Stmt
-	}
-
 	return &ExecStmt{
 		InfoSchema: infoSchema,
 		Plan:       finalPlan,
 		Expensive:  stmtCount(stmtNode, finalPlan, ctx.GetSessionVars().InRestrictedSQL),
 		Cacheable:  plan.Cacheable(stmtNode),
 		Text:       stmtNode.Text(),
-		ReadOnly:   ast.IsReadOnly(readOnlyCheckStmt),
+		// this should not be 'plan.execute'
+		ReadOnly: ast.IsReadOnly(stmtNode),
```

Comment:
It can be `plan.Execute`
There are two types of prepared statement binary and text.